### PR TITLE
Add first step of data processing, ML models examples, and plotting

### PR DESCRIPTION
**Pull request by davidcechak**

### CODE REVIEW

File: download_data.ipynb
Comment (katarinagresova) [inline download_data.ipynb:1]: I would change this jupyter notebook to either python or shell script. There are only shell commands in it. In general, I would keep jupyter notebooks only for initial experimentation and example showcases. But final code will be in python/shell scripts

File: ML_training_and_evaluation.classification.ipynb
Comment (katarinagresova) [inline ML_training_and_evaluation.classification.ipynb:1]: Create a folder `examples/` and put example notebooks there

File: models/__init__.py
Comment (katarinagresova) [inline models/__init__.py:1]: I would create a folder for code, something like `src/` or `src/miRNA/` and put all python modules there.

File: README.md [hunk 1]
```diff
@@ -1,4 +1,32 @@
 ## Scanning the transcriptome for ago2:miRNA to mRNA binding sites and predicting fold change
 
-### Data TargetScan 2015 7miRs HCT116 transfection
-https://drive.google.com/drive/folders/15pX8qrauPJbKYo6r07jRtBLZaRTU6LdO?usp=sharing
+### Set up the environment
+Create a conda env from environment/spec-file.txt by running ```conda create --name YOUR_ENVIRONMENT_NAME --file spec-file.txt```
+
```
Comment (katarinagresova) [inline README.md:6]: Add `conda activate YOUR_ENVIRONMENT_NAME`

File: README.md [hunk 2]
```diff
@@ -1,4 +1,32 @@
...
+### Download the data
+Run the ```download_data.ipynb```
```
Comment (katarinagresova) [inline README.md]: This is not explaining how it should be run. Converting to python/bash script and providing a command to run it would be better.

Comment (katarinagresova) [inline README.md]: When following the README, I am not able to run this notebook - gdown: command not found

File: README.md [hunk 3]
```diff
@@ -1,4 +1,32 @@
...
+
+### Example ML notebooks
+After you download the datasets, you can play with example notebooks ```ML_training_and_evaluation.regression.ipynb``` and ```ML_training_and_evaluation.classification.ipynb```
+
+### Data processing
```
Comment (katarinagresova) [inline README.md]: This is confusing. First, we download data, then we play with models and as the last step, we do data processing? I guess downloaded data are already processed by this pipeline. I would split it into two parts:
1. downloading processed data
2. processing data from scratch

Maybe I wouldn't even put processing data from scratch into README. I would create a folder where are things related to processing pipeline will live. And just mention it in the README

File: models/seeds.py [hunk 4]
```diff
@@ -0,0 +1,180 @@
...
+SEED_TYPES = {
+    'kmer8': seeds_8mer,
+    'kmer7': seeds_7mer,
+    'kmer6': seeds_6mer,
+    'kmer6_bulge': seeds_6mer_bulge,
+    'kmer6_bulge_or_mismatch': seeds_6mer_bulge_or_mismatch
+}
+
+
+def plot_prc_with_seeds(data, seed_types, methods, title=''):
```
Comment (katarinagresova) [inline models/seeds.py]: I would move this to utils/plots.py

Also, it would be nice to make this function more generic. Requiring a dictionary of methods is hard to use. And def do not require `actual` field for every method. It should be inputed only once - it should be the same for every method and here you are risking a typo of using different actuals for different methods. 
If you have seed predictions in `data` dataframe, just add methods as columns to dataframe as well.

File: ML_training_and_evaluation.regression.ipynb
Comment (katarinagresova) [inline ML_training_and_evaluation.regression.ipynb:1]: Create a folder examples/ and put example notebooks there

File: README.md [hunk 5]
```diff
@@ -1,4 +1,32 @@
...
-https://drive.google.com/drive/folders/15pX8qrauPJbKYo6r07jRtBLZaRTU6LdO?usp=sharing
+### Set up the environment
+Create a conda env from environment/spec-file.txt by running ```conda create --name YOUR_ENVIRONMENT_NAME --file spec-file.txt```
+
+### Download the data
+Run the ```download_data.ipynb```
+
+### Example ML notebooks
+After you download the datasets, you can play with example notebooks ```ML_training_and_evaluation.regression.ipynb``` and ```ML_training_and_evaluation.classification.ipynb```
+
```
Comment (katarinagresova) [inline README.md:8]: I would add something like an Evaluation section. If I have my model, how can I compare it to already existing solutions (probably TargetScan in this case)? I imagine a python/bash script that will take for example file with predictions that I produced with my model and it will create plots. 

```python
python evaluate.py --pred_file PATH_TO_FILE_WITH_PREDS --predictors NAMES_OF_COLUMNS_WITH_PREDICTIONS
```

File: data/processed/GRCh37.p13 hg19/README.md
Comment (katarinagresova) [inline data/processed/GRCh37.p13 hg19/README.md:1]: Empty README. And the whole folder is empty. Why is this here? If it is used during data preprocessing, it can be created by pipeline.

File: data/target_scan_8/Gene_info.txt
Comment (katarinagresova) [inline data/target_scan_8/Gene_info.txt:1]: What is this file? Add README

File: data/preprocessed_for_ml/README.md
Comment (katarinagresova) [inline data/preprocessed_for_ml/README.md:1]: This folder can be created when downloading data. Also, description of columns is missing.

File: data/GRCh37.p13 hg19/UCSC/README.md
Comment (katarinagresova) [inline data/GRCh37.p13 hg19/UCSC/README.md:1]: Again, this folder can be created when needed.

File: data/GRCh37.p13 hg19/utr_sequences_hg19.sh
Comment (katarinagresova) [inline data/GRCh37.p13 hg19/utr_sequences_hg19.sh:1]: Why is this script here and not included in `download_data.ipynb` or in proprocessing pipeline?

File: models/constants.py [hunk 6]
```diff
@@ -0,0 +1,125 @@
+SEEDS_TO_COUNT = [
+    'kmer8_count',
+    'kmer7_count',
+    'kmer6_count',
+    'kmer6_bulge_count',
+    'kmer6_mismatch_count',
+]
+
+SEED_COUNT_COLUMNS = [
```
Comment (katarinagresova) [inline models/constants.py]: I think you can define this in a smarter way. You already have `SEEDS_TO_COUNT`. Just define array of counts you want to use. And `SEED_COUNT_COLUMNS` is then combination of `SEEDS_TO_COUNT` and array of counts.

File: models/constants.py [hunk 7]
```diff
@@ -0,0 +1,125 @@
...
+       'kmer7_count_5', 'kmer7_count_6', 'kmer6_count_1', 'kmer6_count_2',
+       'kmer6_count_3', 'kmer6_count_4', 'kmer6_count_5', 'kmer6_count_6',
+       'kmer6_bulge_count_1', 'kmer6_bulge_count_2', 'kmer6_bulge_count_3',
+       'kmer6_bulge_count_4', 'kmer6_bulge_count_5', 'kmer6_bulge_count_6',
+       'kmer6_bulge_or_mismatch_count_1', 'kmer6_bulge_or_mismatch_count_2',
+       'kmer6_bulge_or_mismatch_count_3', 'kmer6_bulge_or_mismatch_count_4',
+       'kmer6_bulge_or_mismatch_count_5', 'kmer6_bulge_or_mismatch_count_6',
+]
+
+TARGETSCAN_COLUMN_TO_SEQUENCE = {
```
Comment (katarinagresova) [inline models/constants.py]: Providing links for all sequences, where you took them form, would be great.

File: models/constants.py [hunk 8]
```diff
@@ -0,0 +1,125 @@
...
+    'hsa-miR-16-5p': 'TAGCAGCACGTAAATATTGGCG', 
+    'hsa-miR-106b-5p': 'TAAAGTGCTGACAGTGCAGAT', 
+    'hsa-miR-200a-3p': 'TAACACTGTCTGGTAACGATGT', # https://mirbase.org/hairpin/MI0000737
+    'hsa-miR-200b-3p': 'TAATACTGCCTGGTAATGATGA', 
+    'hsa-miR-215-5p': 'ATGACCTATGAATTGACAGAC', 
+    'hsa-let-7c-5p': 'TGAGGTAGTAGGTTGTATGGTT', 
+    'hsa-miR-103a-3p': 'AGCAGCATTGTACAGGGCTATGA'
+}
+
+PSILAC_DATA = {
```
Comment (katarinagresova) [inline models/constants.py]: Providing links for all sequences, where you took them form, would be great.

File: models/constants.py [hunk 9]
```diff
@@ -0,0 +1,125 @@
...
+        'name': 'hsa-miR-16-5p',
+    },
+    'log2FC median miR30': {
+        'sequence': 'TGTAAACATCCTCGACTGGAAG',  # hsa-miR-30a-5p
+        'name': 'hsa-miR-30a-5p',  # https://mirbase.org/hairpin/MI0000088
+    },
+}
+
+
+HELA_TRANSFACTION_DATA = {
```
Comment (katarinagresova) [inline models/constants.py]: Providing links for all sequences, where you took them form, would be great.

File: models/seeds.py [hunk 10]
```diff
@@ -0,0 +1,180 @@
...
+    for seq in seeds:
+        seed_counts += gene.count(seq)
+        if seed_counts > thresh:
+            return 1
+        
+    return 0
+
+
+def get_seed_locations(gene, miRNA, get_seeds):
+    # TODO: allow imperfect match
```
Comment (katarinagresova) [inline models/seeds.py]: What do you mean be perfect match?

File: models/seeds.py [hunk 11]
```diff
@@ -0,0 +1,180 @@
...
+    nn = {'A': 'T', 'C': 'G', 'G': 'C', 'T': 'A'}
+    return "".join(nn[n] for n in reversed(st))
+
+
+def classify_site(gene, miRNA, get_seeds, thresh):
+
+    seeds = get_seeds(miRNA)
+    seed_counts = 0
+
+    # TODO: is it possible that one position will be counted as multiple seeds?
```
Comment (katarinagresova) [inline models/seeds.py]: I know this TODO might be from me, but it should be solved.

File: models/seeds.py [hunk 12]
```diff
@@ -0,0 +1,180 @@
...
+    
+    for seq in seeds:
+        start_indexes = [m.start() for m in re.finditer(seq, gene)]
+        for i in start_indexes:
+            seed_loci_flags[i:i+len(seq)] = [1] * len(seq)
+            
+    return seed_loci_flags
+
+
+def get_seed_locations(gene, miRNA, get_seeds):
```
Comment (katarinagresova) [inline models/seeds.py]: Why is this function here 2x?

File: models/seeds.py [hunk 13]
```diff
@@ -0,0 +1,180 @@
...
+        for i in start_indexes:
+            seed_loci_flags[i:i+len(seq)] = [1] * len(seq)
+
+#     if len(seed_loci_flags) == len(gene):
+#         print('2 -OK -> len(seed_loci_flags) == len(gene)')
+        
+    return seed_loci_flags
+
+
+SEED_TYPES = {
```
Comment (katarinagresova) [inline models/seeds.py]: Use better name.

File: utils/data_processing.py [hunk 14]
```diff
@@ -0,0 +1,45 @@
...
+
+
+def get_ids_without_version(id_x):
+    split = id_x.split('.')
+    if len(split) > 1:
+        return split[0]
+    else:
+        return id_x
+
+def parse_biomart_fasta(fasta_path):
```
Comment (katarinagresova) [inline utils/data_processing.py]: Add comment what this function does and why? What is format of input? What is format of output?

File: utils/plots.py [hunk 15]
```diff
@@ -0,0 +1,152 @@
...
+from scipy.stats import spearmanr
+from scipy.stats import pearsonr
+
+
+def plot_multiple_cdfs_with_medians(num_of_top_preds, dataframes_for_top_preds, labels_for_top_preds, lfc_values_list, labels, title_sufix):
+    """
+    Plots the CDFs for multiple sets of log2 fold change (LFC) values with median lines from multiple dataframes.
+    
+    Parameters:
+    - num_of_top_preds: Number of top predictions for each miRNA
```
Comment (katarinagresova) [inline utils/plots.py]: Do not require this param. Give it a default value. And if somebody wants to change it, they can.

File: utils/plots.py [hunk 16]
```diff
@@ -0,0 +1,152 @@
...
+import matplotlib.pyplot as plt
+from sklearn.metrics import r2_score
+from sklearn.feature_selection import r_regression
+from scipy.stats import spearmanr
+from scipy.stats import pearsonr
+
+
+def plot_multiple_cdfs_with_medians(num_of_top_preds, dataframes_for_top_preds, labels_for_top_preds, lfc_values_list, labels, title_sufix):
+    """
+    Plots the CDFs for multiple sets of log2 fold change (LFC) values with median lines from multiple dataframes.
```
Comment (katarinagresova) [inline utils/plots.py]: This is a little complicated plot to interpret. Add a description of what the plot shows.

File: utils/plots.py [hunk 17]
```diff
@@ -0,0 +1,152 @@
...
+    
+    Parameters:
+    - num_of_top_preds: Number of top predictions for each miRNA
+    - dataframes_for_top_preds: List of DataFrames, each containing 'miRNA', predictions (denoted by label), and fold changes (denoted by 'fold_change'). Only top N predictions for miRNA will be plotted
+    - labels_for_top_preds: List of prediction labels
+    - lfc_values_list: List of arrays of LFC values, all values will be plotted
```
Comment (katarinagresova) [inline utils/plots.py]: Why are lfc_values_list and labels as separate input parameters and not part of dataframes_for_top_preds? Maybe it makes sense with connection to other parts, just asking.

File: ML_training_and_evaluation.classification.ipynb
Comment (katarinagresova) [inline ML_training_and_evaluation.classification.ipynb:1]: Unable to run the whole notebook - `NameError: name 'data_train_copy' is not defined`. Make sure example notebooks run whole before publishing them

**Review by katarinagresova — CHANGES_REQUESTED** [review]

File: utils/plots.py [hunk 18]
```diff
@@ -0,0 +1,152 @@
...
+    importance_df = pd.DataFrame({
+        'Feature Names': feature_names,
+        'Importances': feature_importances
+    })
+
+    # Sort by importance
+    importance_df = importance_df.sort_values('Importances', ascending=False)
+
+    # Create the plot
+    plt.figure(figsize=(5, 4),dpi=150)
```
Comment (katarinagresova) [inline utils/plots.py]: For publication, you should do dpi of at least 300

File: ML_training_and_evaluation.classification.ipynb
Comment (davidcechak) [reply ML_training_and_evaluation.classification.ipynb:1]: done

File: README.md (hunk 1 above)
Comment (davidcechak) [reply README.md:6]: done

File: ML_training_and_evaluation.regression.ipynb
Comment (davidcechak) [reply ML_training_and_evaluation.regression.ipynb:1]: done

File: README.md (hunk 3 above)
Comment (davidcechak) [reply README.md]: Good idea! Thanks.
Done

File: README.md (hunk 2 above)
Comment (davidcechak) [reply README.md]: dona

Comment (davidcechak) [reply README.md]: done

File: ML_training_and_evaluation.classification.ipynb
Comment (davidcechak) [reply ML_training_and_evaluation.classification.ipynb:1]: done

File: download_data.ipynb
Comment (davidcechak) [reply download_data.ipynb:1]: done

File: models/seeds.py (hunk 11 above)
Comment (davidcechak) [reply models/seeds.py]: Adding to backlog

File: utils/plots.py (hunk 15 above)
Comment (davidcechak) [reply utils/plots.py]: done

File: utils/plots.py (hunk 17 above)
Comment (davidcechak) [reply utils/plots.py]: The difference is that lfc_values_list FC values are plotted as a whole, whereas for dataframes_for_top_preds we pick the top 16 predictions and plot their FC. 
The lfc_values_list serves for real FC values. 
The dataframes_for_top_preds is for predictions.
It is a generic function where you can plot only one of the groups or their combination. 
I renamed the variables and updated the function description to make it more clear.

File: utils/plots.py (hunk 18 above)
Comment (davidcechak) [reply utils/plots.py]: done

File: utils/plots.py (hunk 16 above)
Comment (davidcechak) [reply utils/plots.py]: done

File: utils/data_processing.py (hunk 14 above)
Comment (davidcechak) [reply utils/data_processing.py]: done

File: models/seeds.py (hunk 12 above)
Comment (davidcechak) [reply models/seeds.py]: done

File: models/constants.py (hunk 6 above)
Comment (davidcechak) [reply models/constants.py]: Good point!
Done

File: models/constants.py (hunk 7 above)
Comment (davidcechak) [reply models/constants.py]: Done

File: models/constants.py (hunk 9 above)
Comment (davidcechak) [reply models/constants.py]: Adding to the backlog, will be solved when adding the other datasets

File: models/constants.py (hunk 8 above)
Comment (davidcechak) [reply models/constants.py]: Adding to the backlog, will be solved when adding the other datasets

File: models/seeds.py (hunk 13 above)
Comment (davidcechak) [reply models/seeds.py]: Done

File: models/seeds.py (hunk 10 above)
Comment (davidcechak) [reply models/seeds.py]: By imperfect I meant the string match. Allow some kind of a loose matching. However, that is not really the right place for that, since here we want to find the specific seed strings as defined. Removing the comment

File: models/seeds.py (hunk 4 above)
Comment (davidcechak) [reply models/seeds.py]: done

File: models/__init__.py
Comment (katarinagresova) [reply models/__init__.py:1]: done

File: ML_training_and_evaluation.classification.ipynb
Comment (katarinagresova) [reply ML_training_and_evaluation.classification.ipynb:1]: done

File: pipelines/process_utr_sequences_hg19.fc.representative_transcript.ipynb
Comment (katarinagresova) [inline pipelines/process_utr_sequences_hg19.fc.representative_transcript.ipynb:1]: Input files are not available
```
IN_UTR_SEQ_TXT_PATH = 'data/GRCh37.p13 hg19/utr_sequences_hg19.txt'
MIRNA_FCS = 'data/fold_change/mirna_fcs.csv'
TS_GENE_INFO = 'data/target_scan_8/Gene_info.txt'
ID_MAP = 'data/GRCh37.p13 hg19/UCSC/id_map'

OUT_UTR_DF_MIRNA_FC_PATH = 'data/processed/GRCh37.p13 hg19/3utr.sequences.refseq_id.mirna_fc.pkl'
```
Not all files are available, nor a way how to download them.

File: src/funmirtar/utils/data_processing.py [hunk 19]
```diff
@@ -0,0 +1,94 @@
+from Bio import SeqIO
+import pandas as pd
+
+
+def get_ids_without_version(id_x):
```
Comment (katarinagresova) [inline src/funmirtar/utils/data_processing.py:5]: What is this function doing? It is not used anywhere.

File: src/funmirtar/utils/data_processing.py [hunk 20]
```diff
@@ -0,0 +1,94 @@
...
+                    "ensembl_transcript_id": parts[12],
+                    "ensembl_transcript_id_version": parts[13],
+
+                    "sequence": record.seq,
+                }
+            )
+    return sequences
+
+
+def get_train_and_test_set(data_folder_path, dataset_name, mirna_names, file_extension_train, file_extension_test):
```
Comment (katarinagresova) [inline src/funmirtar/utils/data_processing.py:51]: What is this function doing? It is not used anywhere.

File: src/funmirtar/utils/plots.py [hunk 21]
```diff
@@ -0,0 +1,200 @@
...
+    title_sufix='', num_of_top_preds=16
+):
+    """
+    Plots the Cumulative Density Function (CDF) of log2 fold change (LFC) values with median lines for multiple sets of data.
+    Each line represents a different dataset, showing the proportion of LFC values below a given threshold.
+    The difference is that values_list_for_all_lfc FC values are plotted as a whole, whereas for dataframes_for_top_preds we pick the top 16 predictions and plot only their LFC.
+    It is possible to plot only one group (top 16 or all LFC values) and leave the other empty.
+
+    Parameters:
+    - dataframes_for_top_preds: List of DataFrames, each containing 'miRNA', predictions (denoted by label), and fold changes (denoted by 'fold_change'). Only top N predictions for each miRNA are picked and their LFC is plotted.
```
Comment (katarinagresova) [inline src/funmirtar/utils/plots.py]: Better describe what columns are expected to be in a DataFrame.

File: src/funmirtar/utils/plots.py [hunk 22]
```diff
@@ -0,0 +1,200 @@
...
+):
+    """
+    Plots the Cumulative Density Function (CDF) of log2 fold change (LFC) values with median lines for multiple sets of data.
+    Each line represents a different dataset, showing the proportion of LFC values below a given threshold.
+    The difference is that values_list_for_all_lfc FC values are plotted as a whole, whereas for dataframes_for_top_preds we pick the top 16 predictions and plot only their LFC.
+    It is possible to plot only one group (top 16 or all LFC values) and leave the other empty.
+
+    Parameters:
+    - dataframes_for_top_preds: List of DataFrames, each containing 'miRNA', predictions (denoted by label), and fold changes (denoted by 'fold_change'). Only top N predictions for each miRNA are picked and their LFC is plotted. 
+    - labels_for_top_preds: List of prediction labels, one label for each method
```
Comment (katarinagresova) [inline src/funmirtar/utils/plots.py]: Is this a ground truth label? It is not clear from description.

File: src/funmirtar/utils/plots.py [hunk 23]
```diff
@@ -0,0 +1,200 @@
...
+    - values_list_for_all_lfc: List of arrays of LFC values, all values will be plotted
```
Comment (katarinagresova) [inline src/funmirtar/utils/plots.py]: is this 1D array? array for each miRNA? unclear

File: src/funmirtar/utils/plots.py [hunk 24]
```diff
@@ -0,0 +1,200 @@
...
+    - labels_for_all_flc: List of labels_for_all_flc for values_list_for_all_lfc, one label for each LFC array
```
Comment (katarinagresova) [inline src/funmirtar/utils/plots.py]: What are labels in this context? Unclear.

File: src/funmirtar/utils/plots.py [hunk 25]
```diff
@@ -0,0 +1,200 @@
...
+    - title_sufix: To be used in plot title
+    - num_of_top_preds: Number of top predictions for each miRNA
```
Comment (katarinagresova) [inline src/funmirtar/utils/plots.py]: Can we select number of top predictions? In description, you are mentioning fixed number 16

**Review by katarinagresova — CHANGES_REQUESTED** [review]

File: src/funmirtar/utils/plots.py (hunk 22 above)
Comment (katarinagresova) [reply src/funmirtar/utils/plots.py]: Changed function input params.

File: src/funmirtar/utils/plots.py (hunk 24 above)
Comment (katarinagresova) [reply src/funmirtar/utils/plots.py]: Changed function input params.

File: src/funmirtar/utils/plots.py (hunk 25 above)
Comment (katarinagresova) [reply src/funmirtar/utils/plots.py]: Changed function input params.

File: src/funmirtar/utils/plots.py (hunk 23 above)
Comment (katarinagresova) [reply src/funmirtar/utils/plots.py]: Changed function input params.

File: pipelines/process_utr_sequences_hg19.fc.representative_transcript.ipynb
Comment (katarinagresova) [reply pipelines/process_utr_sequences_hg19.fc.representative_transcript.ipynb:1]: Now they are in a readme: data/README.md

File: README.md (hunk 5 above)
Comment (davidcechak) [reply README.md:8]: This is now solved by a pipeline for ML ploting and evalutation `ML_plot.classification.inpynb` and `ML_plot.classification.inpynb` in `pipelines/` directory

File: ML_training_and_evaluation.regression.ipynb
Comment (davidcechak) [reply ML_training_and_evaluation.regression.ipynb:1]: done

File: data/processed/GRCh37.p13 hg19/README.md
Comment (davidcechak) [reply data/processed/GRCh37.p13 hg19/README.md:1]: Done

File: data/target_scan_8/Gene_info.txt
Comment (davidcechak) [reply data/target_scan_8/Gene_info.txt:1]: Done

File: data/GRCh37.p13 hg19/UCSC/README.md
Comment (davidcechak) [reply data/GRCh37.p13 hg19/UCSC/README.md:1]: Done

File: data/GRCh37.p13 hg19/utr_sequences_hg19.sh
Comment (davidcechak) [reply data/GRCh37.p13 hg19/utr_sequences_hg19.sh:1]: Script deleted and information about the data source moved to data/readme.md

File: src/funmirtar/utils/data_processing.py (hunk 19 above)
Comment (davidcechak) [reply src/funmirtar/utils/data_processing.py:5]: Used in the other part of data preprocessing. I will leave it here for now, since soon we will push parts of code that use it

File: src/funmirtar/utils/data_processing.py (hunk 20 above)
Comment (davidcechak) [reply src/funmirtar/utils/data_processing.py:51]: Used in the other part of data preprocessing. I will leave it here for now, since soon we will push parts of code that use it

File: data/preprocessed_for_ml/README.md
Comment (davidcechak) [reply data/preprocessed_for_ml/README.md:1]: The folder is now created only when needed.
A task to describe columns written down to backlog.

File: data/README.md [hunk 26]
```diff
@@ -0,0 +1,43 @@
+# Data sources
+
+## data/target_scan_8 (downloaded on 1.6.2024)
```
Comment (katarinagresova) [inline data/README.md]: Header levels are confusing here, they are not consistent.
I would do level 2 description - so in this case it would be "Predicted Targets context++ scores (default predictions) - (17.67 MB)".
If data is already present in the GitHub repository, mention it in the section. Something like:
```markdown
## Predicted Targets context++ scores (default predictions)
Source: [https://www.targetscan.org/cgi-bin/targetscan/data_download.vert80.cgi](https://www.targetscan.org/cgi-bin/targetscan/data_download.vert80.cgi)
Destination: data/target_scan_8
Downloaded on: 1.6.2024
Source update on: 23.5.2022
Size: 17.67 MB
Description: Context++ scores for predicted (conserved) targets of conserved miRNA families
Columns: Gene ID, Gene Symbol, Transcript ID, Species ID, miRNA, Site type, UTR start, UTR end, context++ score, context++ score percentile, weighted context++ score, weighted context++ score percentile, predicted relative KD
Number of rows: 1,397,979
```
Please update description for all sources

Comment (katarinagresova) [inline data/README.md]: I think you are missing name od file here. And I don't see this file in the repo. There is only `Gene_info.txt` and that is described in a different part.

File: data/README.md [hunk 27]
```diff
@@ -0,0 +1,43 @@
+# Data sources
```
Comment (katarinagresova) [inline data/README.md]: I got confused by this file and the whole folder. Is `download_data_for_preprocessing.py` downloading all missing files here? If yes, please mention it somewhere.

File: data/processed/GRCh37.p13 hg19/README.md
Comment (katarinagresova) [inline data/processed/GRCh37.p13 hg19/README.md:1]: Do we need this empty file?

File: pipelines/ML_plot.classification.ipynb
Comment (katarinagresova) [inline pipelines/ML_plot.classification.ipynb:1]: For the future, jupyter notebooks would be only for examples. Pipelines would be scripts. Something you can automatically run and it produces some results.

File: pipelines/ML_prediction.classification.ipynb
Comment (katarinagresova) [inline pipelines/ML_prediction.classification.ipynb:1]: What is the difference between this file and `ML_training_and_evaluation.classification.ipynb`? If there is the same functionality, it could be extracted to some common place.

File: pipelines/ML_prediction.regression.ipynb
Comment (katarinagresova) [inline pipelines/ML_prediction.regression.ipynb:1]: The same as for classification.

File: pipelines/README.md
Comment (katarinagresova) [inline pipelines/README.md:1]: There are more files in the `pipelines` folder as they are described here. Please add every file in this folder into README and briefly describe it. What are inputs? What are outputs?

File: pipelines/add_local_features.ipynb
Comment (katarinagresova) [inline pipelines/add_local_features.ipynb:1]: This could very easily be a python script. Please make it one.

File: pipelines/add_local_features.ipynb [hunk 28]
```diff
@@ -0,0 +1,170 @@
...
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "id": "98bdf818-9377-4360-b8db-1fcada81af48",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "import numpy as np\n",
```
Comment (katarinagresova) [inline pipelines/add_local_features.ipynb]: Most of the inputs are unused. Remove them.

File: README.md [hunk 29]
```diff
@@ -1,4 +1,16 @@
...
-### Data TargetScan 2015 7miRs HCT116 transfection
-https://drive.google.com/drive/folders/15pX8qrauPJbKYo6r07jRtBLZaRTU6LdO?usp=sharing
+## Running example models and evaluation
+### Set up the environment
+Create a conda env by running ```conda create --name YOUR_ENVIRONMENT_NAME --file spec-file.txt```
+
+Activate an existing environment by ```conda activate YOUR_ENVIRONMENT_NAME```
+
+### Download the data
+Run ```./data/download_data_for_ML.sh``` which installs the ```gdown``` package (if not installed yet) through pip and downloads the datasets.
```
Comment (katarinagresova) [inline README.md:10]: There is no `./data/download_data_for_ML.sh` script.

File: pipelines/create_seeds_prediction_file.ipynb
Comment (katarinagresova) [inline pipelines/create_seeds_prediction_file.ipynb:1]: Again, very easy to make it into python script. Input and output parameters passed as a command line arguments.

File: pipelines/create_target_scan_prediction_file.ipynb
Comment (katarinagresova) [inline pipelines/create_target_scan_prediction_file.ipynb:1]: Again, very easy to make it into python script. Input and output parameters passed as a command line arguments.

File: data/README.md (hunk 26 above)
Comment (davidcechak) [reply data/README.md]: I see the confusion. This line is supposed to inform about the whole folder. The specific files are then described under this title. However, it is a confusing structure, I will rewrite it based on your other [comment](https://github.com/ML-Bioinfo-CEITEC/miRNA/pull/6#discussion_r1692777983)

File: data/processed/GRCh37.p13 hg19/README.md
Comment (davidcechak) [reply data/processed/GRCh37.p13 hg19/README.md:1]: Removed

File: README.md (hunk 29 above)
Comment (davidcechak) [reply README.md:10]: Moved

File: pipelines/add_local_features.ipynb (hunk 28 above)
Comment (davidcechak) [reply pipelines/add_local_features.ipynb]: Removed

File: pipelines/add_local_features.ipynb
Comment (davidcechak) [reply pipelines/add_local_features.ipynb:1]: That will be a part of the next feature pull request

File: pipelines/create_seeds_prediction_file.ipynb
Comment (davidcechak) [reply pipelines/create_seeds_prediction_file.ipynb:1]: That will be a part of the next feature pull request

File: pipelines/create_target_scan_prediction_file.ipynb
Comment (davidcechak) [reply pipelines/create_target_scan_prediction_file.ipynb:1]: That will be a part of the next feature pull request

File: pipelines/ML_prediction.classification.ipynb
Comment (davidcechak) [reply pipelines/ML_prediction.classification.ipynb:1]: This file will be turned into a pipeline module script in the next pull request. `ML_training_and_evaluation.classification.ipynb` is an example notebook. It should be refactored once we have the modular script and the pipeline approach done

File: pipelines/ML_prediction.regression.ipynb
Comment (davidcechak) [reply pipelines/ML_prediction.regression.ipynb:1]: Same explanation as for the classification.
This file will be turned into a pipeline module script in the next pull request. `ML_training_and_evaluation.regression.ipynb` is an example notebook. It should be refactored once we have the modular script and the pipeline approach done

File: pipelines/README.md
Comment (davidcechak) [reply pipelines/README.md:1]: That will be a part of the next feature pull request

File: pipelines/ML_plot.classification.ipynb
Comment (davidcechak) [reply pipelines/ML_plot.classification.ipynb:1]: That will be a part of the next feature pull request

File: data/README.md (hunk 26 above)
Comment (davidcechak) [reply data/README.md]: Done

File: data/README.md (hunk 27 above)
Comment (davidcechak) [reply data/README.md]: Described in read me

**Review by katarinagresova — APPROVED** [review]
